node/src/types/block/hot_block.rs

```diff
@@ -0,0 +1,170 @@
+            && self.execution_results != other.execution_results
+        {
+            return Err(MergeMismatchError::ExecutionResults);
+        }
+
+        self.state = self.state.merge(other.state)?;
+
+        Ok(self)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use rand::Rng;
+
+    use casper_types::testing::TestRng;
+
+    use super::*;
+    use crate::types::Deploy;
+
+    #[test]
+    fn should_merge_when_same_non_empty_execution_results() {
+        let mut rng = TestRng::new();
+
+        let block = Arc::new(Block::random(&mut rng));
+        let deploy = Deploy::random(&mut rng);
+        let execution_results = vec![(*deploy.hash(), deploy.take_header(), rng.gen())];
+        let state = State::new_synced();
+
+        let hot_block1 = HotBlock::new(Arc::clone(&block), execution_results.clone(), state);
+        let hot_block2 = HotBlock::new(Arc::clone(&block), execution_results.clone(), state);
+
+        let merged = hot_block1.clone().merge(hot_block2.clone()).unwrap();
+
+        assert_eq!(merged.block, block);
+        assert_eq!(merged.execution_results, execution_results);
+        assert_eq!(merged.state, State::new_synced());
+        assert_eq!(hot_block2.merge(hot_block1).unwrap(), merged)
+    }
+
+    #[test]
+    fn should_merge_when_both_empty_execution_results() {
+        let mut rng = TestRng::new();
+
+        let block = Arc::new(Block::random(&mut rng));
+        let state = State::new();
+
+        let hot_block1 = HotBlock::new(Arc::clone(&block), vec![], state);
+        let hot_block2 = HotBlock::new(Arc::clone(&block), vec![], state);
+
+        let merged = hot_block1.clone().merge(hot_block2.clone()).unwrap();
+
+        assert_eq!(merged.block, block);
+        assert!(merged.execution_results.is_empty());
+        assert_eq!(merged.state, state);
+        assert_eq!(hot_block2.merge(hot_block1).unwrap(), merged)
+    }
+
+    #[test]
+    fn should_merge_when_one_empty_execution_results() {
+        let mut rng = TestRng::new();
+
+        let block = Arc::new(Block::random(&mut rng));
+        let deploy = Deploy::random(&mut rng);
+        let execution_results = vec![(*deploy.hash(), deploy.take_header(), rng.gen())];
+        let state = State::new_immediate_switch();
+
+        let hot_block1 = HotBlock::new(Arc::clone(&block), execution_results.clone(), state);
+        let hot_block2 = HotBlock::new(Arc::clone(&block), vec![], state);
+
+        let merged = hot_block1.clone().merge(hot_block2.clone()).unwrap();
+
+        assert_eq!(merged.block, block);
+        assert_eq!(merged.execution_results, execution_results);
+        assert_eq!(merged.state, state);
+        assert_eq!(hot_block2.merge(hot_block1).unwrap(), merged)
+    }
+
+    #[test]
+    fn should_fail_to_merge_different_blocks() {
+        let mut rng = TestRng::new();
+
+        let block1 = Arc::new(Block::random(&mut rng));
+        let block2 = Arc::new(Block::random(&mut rng));
```

Comment:
Unlikely, but I think it may lead to test flakiness similar to the one fixed [here](https://github.com/casper-network/casper-node/pull/3545/commits/c69fbd287a7646eefa689b92e58f7f6e5aae6ce8).

Reply:
Good point!

Fixed in [e4d8cc6](https://github.com/casper-network/casper-node/pull/3549/commits/e4d8cc609a03671e5923825bf06174f6c64e9e7f).